src/mem/mem_scheduler.hh

```diff
@@ -0,0 +1,290 @@
+
+    /**
+     * Port on the memory-side that receives responses.
+     * Mostly just forwards requests to the owner
+     */
+    class MemSidePort : public RequestPort
+    {
+      private:
+        /// The object that owns this object (MemScheduler)
+        MemScheduler *owner;
+        bool blocked;
+        /// If we tried to send a packet and it was blocked, store it here
+        PacketPtr blockedPacket;
+
+      public:
+        /**
+         * Constructor. Just calls the superclass constructor.
+         */
+        MemSidePort(const std::string& name, MemScheduler *owner) :
+            RequestPort(name, owner), owner(owner), blocked(false), blockedPacket(nullptr)
+        {}
+
+        /**
+         * Send a packet across this port. This is called by the owner and
+         * all of the flow control is hanled in this function.
+         *
+         * @param packet to send.
+         */
+        void sendPacket(PacketPtr pkt);
+
+        bool getBlocked(void);
+
+      protected:
+        /**
+         * Receive a timing response from the response port.
+         */
+        bool recvTimingResp(PacketPtr pkt) override;
+
+        /**
+         * Called by the response port if sendTimingReq was called on this
+         * request port (causing recvTimingReq to be called on the responder
+         * port) and was unsuccesful.
+         */
+        void recvReqRetry() override;
+
+        /**
+         * Called to receive an address range change from the peer responder
+         * port. The default implementation ignores the change and does
+         * nothing. Override this function in a derived class if the owner
+         * needs to be aware of the address ranges, e.g. in an
+         * interconnect component like a bus.
+         */
+        void recvRangeChange() override;
+    };
+
+    /**
+     * Handle the request from the CPU side
+     *
+     * @param requesting packet
+     * @return true if we can handle the request this cycle, false if the
+     *         requestor needs to retry later
+     */
+    bool handleRequest(CPUSidePort *port, PacketPtr pkt);
+
+    /**
+     * Handle the respone from the memory side
+     *
+     * @param responding packet
+     * @return true if we can handle the response this cycle, false if the
+     *         responder needs to retry later
+     */
+    bool handleResponse(PacketPtr pkt);
+
+    /**
+     * Handle a packet functionally. Update the data on a write and get the
+     * data on a read.
+     *
+     * @param packet to functionally handle
+     */
+    void handleFunctional(PacketPtr pkt);
+
+    /**
+     * Return the address ranges this memobj is responsible for. Just use the
+     * same as the next upper level of the hierarchy.
+     *
+     * @return the address ranges this memobj is responsible for
+     */
+    AddrRangeList getAddrRanges() const;
+
+    /**
+     * Tell the CPU side to ask for our memory ranges.
+     */
+    void sendRangeChange();
+
+    void wakeUp();
```

Comment:
What does this function do?

Reply:
When all of the memory side ports are blocked and all the queues in the memory scheduler are full then the nextReqEvent won't be scheduled. We need a wakeup function to schedule the nextReqEvent when memory side ports are no longer blocked.